pkgs/development/cuda-modules/new-modules/package-sets.nix

```diff
@@ -0,0 +1,253 @@
+      # in the lib directory. So long as one of the versions in cudaVersionsInLib is a prefix of the current CUDA
+      # version, we accept the package. We should have a more stringent version check, but no one has written
+      # a sidecar file mapping releases to supported CUDA versions.
+      isSupportedPlatform && hasSupportedCudaVersionInLib
+    else
+      # TODO:
+      isSupportedPlatform && (isDesiredCudaVariant || hasSupportedCudaVersionInLib);
+
+  # Function to update the package set with the new package.
+  updatePackages =
+    meta: packages: package:
+    let
+      # Non-CUDA redist packages, like cuDNN, are multiplexed so we add a suffix to the attribute name.
+      # CUDA redist packages are always the same as the package name.
+      attributeName =
+        if meta.redistName != "cuda" then
+          config.utils.mkVersionedPackageName meta.packageName (majorMinor meta.releaseInfo.version)
+        else
+          meta.packageName;
+
+      # Booleans which instruct us on how to update the package set at this step.
+      packagesHasEntry = packages ? ${attributeName};
+      entryIsOlder = versionOlder packages.${attributeName}.version package.version;
+      packagesHasDefault = packages ? ${meta.packageName};
+      defaultIsOlder = versionOlder packages.${meta.packageName}.version package.version;
+    in
+    packages
+    # We keep point-releases for non-CUDA redist packages because NVIDIA may not release a new point release for
+    # all platforms. For this reason, when we update the package within the package set, we need to check if the
+    # point-release we're currently processing is newer than the one we have in the package set already.
+    # NOTE: We cannot use optionalAttrs here because it causes an infinite recursion error.
+    # NOTE: In the case we're processing a CUDA redistributable, the attribute name and the package name are the same,
+    # so we're effectively replacing the package twice.
+    // {
+      ${attributeName} =
+        if (!packagesHasEntry || entryIsOlder) then package else packages.${attributeName};
+    }
+    // {
+      ${meta.packageName} =
+        if (!packagesHasDefault || defaultIsOlder) then package else packages.${meta.packageName};
+    };
+
+  packageSetBuilder = fullCudaVersion: {
+    name = config.utils.mkVersionedPackageName "cudaPackages" (majorMinor fullCudaVersion);
+    value = makeScope newScope (
+      final:
+      let
+        flattenedIndexFoldFn =
+          packages: meta:
+          let
+            package = pipe ../generic-builders/manifest.nix [
+              # Import the builder function.
+              (flip builtins.import (
+                meta
+                // {
+                  src = fetchzip {
+                    hash = meta.narHash;
+                    url = config.utils.mkRedistURL meta.redistName (config.utils.mkRelativePath meta);
+                  };
```

Comment:
One itch I have about this is that we don't include a proof that these narHash-es are actually derived from the tarballs and hashes declared in upstream manifests.

On the other hand with two layers of FODs, `let a = fetchurl { hash = upstreamHash; ... }; in runCommand "unzipped" { hash = narHash; ... } "tar -xf ${a} ..."`, we could just cache the output of the latter, but we could also rebuild it as long as nvidia keeps providing the tarball, thereby offering a "proof" that the hash is legit... of course this is all rather superficial

Reply:
I'm using that pattern now:

https://github.com/NixOS/nixpkgs/blob/2fc79cf094ca3800a05fe8356278b04fc5d52756/pkgs/development/cuda-modules/new-modules/package-sets.nix#L178-L201

Seems to be working okay?